File: block.js
var Node = require("./node");
var $ = require("./$");

module.exports = class BLOCK extends Node {
  constructor() {
    super();
    this.statements = [];
  }

  run() {
    let list = this.statements;
    this.statements = [];
    return list;
  }

  stage(instruction) {
    if (
      instruction.run &&
      !(instruction.statement instanceof $) &&
      instruction.statement.type != "DECLARATION"
    ) {
      let statement = instruction.statement;
      let block = statement.block;

      if (block && statement.id) {
        block.statements = block.statements.filter(s => s.id != statement.id);
      }

      this.statements.push(instruction.statement);
    }
  }
};

File: src/nucleoid/nuc/BLOCK.py
"""
BLOCK runtime node class.
"""
from typing import Any, List
from .NODE import NODE


class BLOCK(NODE):
    """Block statement node containing multiple statements."""

    def __init__(self, key: Any = None) -> None:
        """
        Initialize a BLOCK instance.

        Args:
            key: The block key identifier
        """
        super().__init__(key)
        self.statements: List[Any] = []
        self.skip: bool = False

    def run(self, scope: Any) -> List[Any]:
        """
        Execute the block statements.

        Args:
            scope: The execution scope

        Returns:
            List of Instruction instances
        """
        # Lazy import to avoid circular dependencies
        from ..Instruction import Instruction

        return [
            Instruction(scope, statement, None, None, None, None)
            for statement in self.statements
        ]
